radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/proxy/ProxyCertificateManager.java

```diff
@@ -0,0 +1,309 @@
+import com.radixdlt.network.p2p.PeerEvent;
+import com.radixdlt.network.p2p.PeersView;
+import com.radixdlt.network.p2p.addressbook.AddressBook;
+import com.radixdlt.networks.Addressing;
+import com.radixdlt.networks.NetworkId;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.radix.time.Time;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Manages the proxy certificates:
+ *  - issued by this node
+ *  - granted to this node
+ *  - received from other peers
+ */
+public final class ProxyCertificateManager {
+	private static final Logger log = LogManager.getLogger();
+
+	private final NodeId self;
+	private final P2PConfig config;
+	private final int networkId;
+	private final Addressing addressing;
+	private final ECKeyOps ecKeyOps;
+	private final PeersView peersView;
+	private final PeerControl peerControl;
+	private final AddressBook addressBook;
+	private final RemoteEventDispatcher<GrantedProxyCertificate> grantedProxyCertificateDispatcher;
+	private final RemoteEventDispatcher<ProxyCertificatesAnnouncement> proxyCertificatesAnnouncementDispatcher;
+	private final Object lock = new Object();
+
+	// the certificates that were granted to this node
+	private final Map<NodeId, ProxyCertificate> receivedProxyCertificates = new ConcurrentHashMap<>();
+
+	// the certificates issued by this node
+	private final Map<NodeId, ProxyCertificate> issuedProxyCertificates = new ConcurrentHashMap<>();
+
+	// the certificates received from other nodes (granted to them)
+	private final Map<NodeId, Set<VerifiedProxyCertificate>> verifiedProxies = new ConcurrentHashMap<>();
+
+	@Inject
+	public ProxyCertificateManager(
+		@Self NodeId self,
+		P2PConfig config,
+		@NetworkId int networkId,
+		Addressing addressing,
+		ECKeyOps ecKeyOps,
+		PeersView peersView,
+		PeerControl peerControl,
+		AddressBook addressBook,
+		RemoteEventDispatcher<GrantedProxyCertificate> grantedProxyCertificateDispatcher,
+		RemoteEventDispatcher<ProxyCertificatesAnnouncement> proxyCertificatesAnnouncementDispatcher
+	) {
+		this.self = Objects.requireNonNull(self);
+		this.config = Objects.requireNonNull(config);
+		this.networkId = networkId;
+		this.addressing = Objects.requireNonNull(addressing);
+		this.ecKeyOps = Objects.requireNonNull(ecKeyOps);
+		this.peersView = Objects.requireNonNull(peersView);
+		this.peerControl = Objects.requireNonNull(peerControl);
+		this.addressBook = Objects.requireNonNull(addressBook);
+		this.grantedProxyCertificateDispatcher = Objects.requireNonNull(grantedProxyCertificateDispatcher);
+		this.proxyCertificatesAnnouncementDispatcher = Objects.requireNonNull(proxyCertificatesAnnouncementDispatcher);
+	}
+
+	public ImmutableSet<ProxyCertificate> getReceivedProxyCertificates() {
+		synchronized (lock) {
```

Comment:
`lock` here necessary given it's a `ConcurrentHashMap`?

Reply:
good catch, no longer needed